### DISCUSSION TITLE
単体テストの考え方/使い方 5.1.5〜5.3.1

### DISCUSSION THREAD

**Discussion by dameleon** [atman-inc/training#99] [General]
- 5.1.5: モックとスタブがそれぞれどのようにコマンドとクエリに結びつくのか？
〜(真ん中省略)
- 5.3.1: ヘキサゴナル・アーキテクチャとは？

行けるとこまで。はみ出たら次週。

上記辺りののディスカッション。
不明点、相談点、確認点、勉強になったポイントなどは各自ディスカッションのコメントをお願いします。

**Comment by koushisa** [comment]
少々難しい概念へ入ってきましたね。

人ごとにコメントが分散していると拾いづらい気がするので、トピックごとにスレッドを作って連ねていくのどうですか。
俯瞰しやすくなりますかね？

以下に叩きつくります。
叩きに合わせるかは任せますが、乗っかる人がいたら乗っかっていただけると、

**Comment by koushisa** [comment]
## 5.1.5

**Reply by koushisa** [reply]
章に対して人ごとにreplyを追加していくイメージ
例えば、ここのreplyは5.1.5に対するコメントだけにする。

章ごとに各自の意見が揃った状態で俯瞰できるため、コミュニケーションの活性化が見込めるのではないかと思います。

(輪読会なので)どんなコメントするかも自由ですが、コメントがない人はまとめきれていない or 意見がないということで。

私も読んだら、このcommentを編集して5.1.5のコメント書いていきます。

---

- 副作用(side effect)の取り扱いは鬼門だと思う
  - 副作用起こすなら戻り値返すな
  - 戻り値返すなら副作用起こすな
- 本当に難しいので自分はこれで過去に関数型プログラミングの沼にはまった(effect systemとかmonadとか)
- FE文脈
  - CQS(≒CQRS)を厳格に実装したReduxも忠実に守っている
  - その結果thunkやsagaみたいな概念が難しく初学者に受け入れられなかったという...
  - FEでもCQSが必要な理由, 経緯等はReduxのスタイルガイドを読んでみると示唆が得られた記憶
  - https://redux.js.org/style-guide/

**Reply by xxManatoxx** [reply]
### コマンド・クエリ分離の原則
- コマンド
  - 戻り値がなく副作用を起こすメソッドを指す
  - テストダブルがモックに該当する
- クエリ
  - コマンドの逆
  - 戻り値を返し、副作用を起こさないメソッドを指す
  - テストダブルがスタブに該当する

- 例外もあるが、基本的には原則に従った方が良い

### 学びや感想
- 戻り値を返すのか、副作用を起こすのかを分離することでコードの可読性が上がる
  - テストで見た時もスタブとモックと明確にダブルが作成しやすくなるため、良い実装だと言えるようになるのかなと感じた
  - 逆にスタブなのかモックなのかが明確に分けられない場合は実装を見直すきっかけになりそう

**Reply by y-kaneda** [reply]
### 5.1.5 モックとスタブがそれぞれどのようにコマンドとクエリに結びつくのか？
- コマンド・クエリ分離(Command Query Separation: CQS)の原則
  - コマンド
    - 戻り値のない、副作用を起こすメソッド
    - モックに紐づく
  - クエリ
    - 副作用は起こさず、何らかの値を返すメソッド
    - スタブに紐づく

**Comment by koushisa** [comment]
## 5.2.0 ~ 5.2.1

**Reply by koushisa** [reply]
### テストではHowではなく、Whatを検証する
- 実装でも同じことが言えるはずで、WhatとHowを分離した関数設計が出来ていないと脆くなりがち
- FE視点だと
  - コンポーネント(What=プレゼンテーション)と状態遷移(How=ビジネスロジック)は切り離すとメンテナンスしやすい
  - →見た目だけの変更は特定のスコープに閉じさせる

### 観察可能な振る舞い = "操作" | "状態"


> ここで心に留めておいてほしいことは、コードが観察可能な振る舞いなのか否かの判断は、クライアントは何なのか、そして、そのクライアントが目標としていることは何なのか、ということによって変わることです。

とてもいい言葉。

**Reply by nmurakami0** [reply]
状態でいうと、フラグはobservable behaviorなのかどうか

UIで使っていれば(e.g., 使用可能フラグ)該当するが、内部的な制御で使っている場合(e.g., 販売に対して支払い済みかのフラグ)は該当しない？

**Reply by y-kaneda** [reply]
## 5.2 観察可能な振る舞い(observable behavior)と実装の詳細(implementation detail)

- テストの壊れやすさに影響を与えるのがリファクタリングへの耐性
- 単体テストがE2Eの領域にまではみ出すとリファクタリングへの耐性は上がるが、保守のしやすさが失われる
- テスト・ケースが実装の詳細と結び付くとfalse positive を生み出してしまう
  - 最終的な振る舞いの結果だけを検証するようにする(ブラックボックス)

### 5.2.1 観察可能な振る舞いと公開されたAPIとの違い
- class や method 等の private, public などのスコープの話
- 操作・状態、オブジェクト指向の話っぽい
- テスト・コードとプロダクション・コードが違うクラス、パッケージになっていて変数や関数のスコープが適切であれば実装の詳細に近づきすぎることを防げる

**Reply by xxManatoxx** [reply]
### プロダクションコードを以下2つの観点から分類
- 公開されたAPIか、プライベートなAPIか
- 観測可能な振る舞いか、実装の詳細か

### 実装の詳細
- 以下に該当しない物が実装の詳細と言える
  - クライアントが目標を達成するために使う公開された**操作**
  - クライアントが目標を達成するために使う公開された**状態**
- コードが観測可能な振る舞いか否かは以下によって変わる
  - クライアントが何なのか
  - クライアントの目標は何なのか

### 学びや感想
- 雰囲気やなんとなく理解できているような気がするがもう何回か読み直してから進みたい
- 現時点での言語化
  - 何でも処理を公開するべきではなく、あくまでクライアントが目標としている事を叶える物を公開
  - 目標を叶えるための詳細な処理は外部に公開せず、あくまで内部的にするべき
- 実装の詳細か、観測可能な振る舞いかの判別を、実際にコードを書く時に意識して考えたいと思いました

**Comment by koushisa** [comment]
## 5.2.2

**Reply by y-kaneda** [reply]
### 5.2.2 公開されたAPIから漏洩する実装の詳細
- カプセル化ちゃんとすれば良い
- 判断基準の一つに、外部から何か一つの結果を達成するために同クラスのメソッドを複数呼んでしまっているかというのがある

**Reply by koushisa** [reply]
> 現実的なプロジェクトとして考えた場合、getアクセサを介してユーザーの現時点の名前を取得することは大いに有り得ることだと考えられます。ここでは、そのような別のユース・ケースがあると想定して〜

よく分かる。
参照系はユースケースと密結合な処理が多く、局所的な処理が必要になりがち。
API提供側が参照系を強く意識すると利用者との結合が強くなってしまう。

**Reply by nmurakami0** [reply]
NormalizeNameがもうちょっと複雑で複数のpublicメソッドから呼ばれていたら、単体でテスト書きたくなりそうだが、この本としてはNameのsetterの内部挙動としてsetterをテストすべきとしている。

**Comment by koushisa** [comment]
## 5.2.3

**Reply by y-kaneda** [reply]
### 5.2.3 きちんと設計されたAPIとカプセル化
- カプセル化は不変条件(invariant)の侵害からコードを守るための手段
  - 勝手に外から呼ばれて変えられてしまったり、逆に外から呼ばれなかったことにより条件を満たさなくなってしまったりする

**Reply by nmurakami0** [reply]
> Tell, Dont' ask
> データを操作するのにメソッドを経由させること

リファクタリングへの耐性としても重要と思われる

**Comment by koushisa** [comment]
## 5.2.4

**Reply by y-kaneda** [reply]
### 5.2.4 状態からの実装の詳細の漏洩
- 変数においても内部でしか利用しないものは private 化して外部から観察されないようにする

**Reply by nmurakami0** [reply]
public, privateが言語仕様として用意されていない言語はどういう意図なのかなとふと思いました

**Comment by koushisa** [comment]
## 5.3.0 ~ 5.3.1

**Reply by y-kaneda** [reply]
### 5.3.1 ヘキサゴナル・アーキテクチャとは？
- オニオン・アーキテクチャ的な話
- 六角形である必要はよく分からないが、一辺が他のアプリケーションとの接続みたいなのを表したい？
- 実装の詳細が隠れたAPIの構造になっていると、同一の目標を達成するための操作に対して外側から中心側までの概要から詳細の粒度の異なる複数のテストになる
- 各々のコードを呼び出したい各クライアントの立ち位置によって観察する対象が変わってくる

**Reply by nmurakami0** [reply]
六角形だと接続面を表しやすいというのは今まで"ヘキサゴナル"を全然気にしていなかったが、なるほどと思いました

**Comment by koushisa** [comment]
## 5.3.2

**Comment by y-kaneda** [comment]
## 5.2 観察可能な振る舞い(observable behavior)と実装の詳細(implementation detail)

- テストの壊れやすさに影響を与えるのがリファクタリングへの耐性
- 単体テストがE2Eの領域にまではみ出すとリファクタリングへの耐性は上がるが、保守のしやすさが失われる
- テスト・ケースが実装の詳細と結び付くとfalse positive を生み出してしまう
  - 最終的な振る舞いの結果だけを検証するようにする(ブラックボックス)

### 5.2.1 観察可能な振る舞いと公開されたAPIとの違い
- class や method 等の private, public などのスコープの話
- 操作・状態、オブジェクト指向の話っぽい
- テスト・コードとプロダクション・コードが違うクラス、パッケージになっていて変数や関数のスコープが適切であれば実装の詳細に近づきすぎることを防げる

### 5.2.2 公開されたAPIから漏洩する実装の詳細
- カプセル化ちゃんとすれば良い
- 判断基準の一つに、外部から何か一つの結果を達成するために同クラスのメソッドを複数呼んでしまっているかというのがある

### 5.2.3 きちんと設計されたAPIとカプセル化
- カプセル化は不変条件(invariant)の侵害からコードを守るための手段
  - 勝手に外から呼ばれて変えられてしまったり、逆に外から呼ばれなかったことにより条件を満たさなくなってしまったりする

### 5.2.4 状態からの実装の詳細の漏洩
- 変数においても内部でしか利用しないものは private 化して外部から観察されないようにする

## 5.3 モックの利用とテストの壊れやすさとの関係

### 5.3.1 ヘキサゴナル・アーキテクチャとは？
- オニオン・アーキテクチャ的な話
- 六角形である必要はよく分からないが、一辺が他のアプリケーションとの接続みたいなのを表したい？
- 実装の詳細が隠れたAPIの構造になっていると、同一の目標を達成するための操作に対して外側から中心側までの概要から詳細の粒度の異なる複数のテストになる
- 各々のコードを呼び出したい各クライアントの立ち位置によって観察する対象が変わってくる

**Comment by myamamoto111** [comment]
## 5.2 観察可能な振る舞いと実装の詳細
### 5.2.1 観察可能な振る舞いと公開されたAPIとの違い
- API
  - 公開されたAPI
  - プライベートなAPI
- コード
  - 観察可能な振る舞い
    - 操作
      - メソッド
    - 状態
      - 現時点でのコンディション
  - 実装の詳細
    - 上記に該当しないもの
理想なのは、システムが公開しているAPIが観察可能な振る舞いと一致し、
システムの全ての実装の詳細がクライアントから完全に隠れるようになっていること。

### 5.2.2 公開されたAPIから漏洩する実装の詳細
理想とすべきAPIの設計は
いかなる目標であれ、１つの操作で目標を達成できるようにすること。

### 5.2.3 きちんと設計されたAPIとカプセル化
- カプセル化
  - 不変条件（常に真(true)の条件）からコードを守るための手段
- 「尋ねるな、命じよ」
  - データを操作するメソッドを用いてデータを構築すること

### 5.2.4 状態からの実装の詳細の漏洩
- APIをきちんと設計すれば、単体テストは自然と質の良いものになる。
- 公開する操作と状態は最小限に抑える。

## 5.3 モックの利用とテストの壊れやすさとの関係
### 5.3.1 ヘキサゴナル・アーキテクチャとは？
- ドメイン層（ビジネス・ロジック）
- アプリケーション・サービス層
- ドメイン層とアプリケーション・サービス層との関心の分離
    - ドメイン層はビジネスロジック
    - アプリケーション・サービス層は、ユース・ケースを集めたもの
- アプリケーション内でのコミュニケーション
    - 依存の流れがアプリケーション・サービス層からドメイン層への一方向
- 外部アプリケーションとのコミュニケーション
    - アプリケーション・サービス層にある共通のインターフェースを介して行われる

書かれている理想はわかるけど、実現可能なのだろうか？
